gateway/src/lib.rs

```diff
@@ -1,10 +1,246 @@
+
+    #[endpoint(callContract)]
+    fn call_contract(
+        &self,
+        destination_chain: ManagedBuffer,
+        destination_contract_address: ManagedBuffer,
+        payload: ManagedBuffer,
+    ) {
+        let caller = self.blockchain().get_caller();
+
+        self.contract_call_event(
+            caller,
+            destination_chain,
+            destination_contract_address,
+            ContractCallData {
+                hash: self.crypto().keccak256(&payload),
+                payload,
+            },
+        );
+    }
+
+    // Can only be called by the appropriate contract address
+    #[endpoint(validateContractCall)]
+    fn validate_contract_call(
+        &self,
+        command_id: &ManagedByteArray<KECCAK256_RESULT_LEN>,
+        source_chain: &ManagedBuffer,
+        source_address: &ManagedBuffer,
+        payload_hash: &ManagedByteArray<KECCAK256_RESULT_LEN>,
+    ) -> bool {
+        let contract_address = &self.blockchain().get_caller();
+
+        let hash = self.get_is_contract_call_approved_key(
+            command_id,
+            source_chain,
+            source_address,
+            contract_address,
+            payload_hash,
+        );
+
+        let valid = self.contract_call_approved().contains(&hash);
+
+        if valid {
+            self.contract_call_approved().remove(&hash);
+
+            self.contract_call_executed_event(command_id);
+        }
+
+        valid
+    }
+
+    // External Functions
+
+    #[endpoint(execute)]
+    fn execute(&self, input: ExecuteInput<Self::Api>) {
+        let message_hash = self.get_message_hash(&input.data);
+
+        let mut allow_operatorship_transfer: bool =
+            self.auth_validate_proof(&message_hash, &input.proof);
+
+        let execute_data: ExecuteData<Self::Api> =
+            ExecuteData::<Self::Api>::top_decode(input.data).unwrap();
+
+        require!(
+            execute_data.chain_id == self.chain_id().get(),
+            "Invalid chain id"
+        );
+
+        let commands_length = execute_data.command_ids.len();
+
+        require!(
+            commands_length == execute_data.commands.len()
+                && commands_length == execute_data.params.len(),
+            "Invalid commands"
+        );
+
+        let selector_approve_contract_call =
+            &ManagedBuffer::new_from_bytes(SELECTOR_APPROVE_CONTRACT_CALL);
+        let selector_transfer_operatorship =
+            &ManagedBuffer::new_from_bytes(SELECTOR_TRANSFER_OPERATORSHIP);
+
+        let command_executed_mapper = self.command_executed();
+        for index in 0..commands_length {
+            let command_id_ref = execute_data.command_ids.get(index);
+            let command_id = command_id_ref.deref();
+
+            if command_executed_mapper.contains(command_id) {
+                continue;
+            }
+
+            let command_ref = execute_data.commands.get(index);
+            let command = command_ref.deref();
+
+            let success: bool;
```

Comment:
I guess this is initialized with false by default. Better write the whole statement (`success = false`)

Reply:
Actually this wasn't even needed (leftover from a previous implementation) so I removed it.